File: bindings/python/bench/bench.py
# CP18 part B: Python binding raw render + repeated/server render workload.
import os, sys, time
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
from nift import Context, Engine

e = Engine.new()
e.set_root("/")
e.set_string("site", "nift")
page = "<p>$[site]</p>"
tpl = "<main>@content</main>"
n = 50000
rounds = 3
r = e.render_sources(page, tpl)  # warm-up (unreported)
if not r.ok:
    raise SystemExit(r.error)
raw_samples = []
req_samples = []
for _ in range(rounds):
    start = time.perf_counter()
    for _ in range(n):  # raw: no request Context, engine-default binding
        r = e.render_sources(page, tpl)
        if not r.ok:
            raise SystemExit(r.error)
    raw_samples.append((time.perf_counter() - start) * 1e9 / n)
    start = time.perf_counter()
    for _ in range(1000):  # request-loop: fresh Context per request
        c = Context()
        c.set_string("who", "w")
        r = e.render_sources(page, tpl, c)
        if not r.ok:
            raise SystemExit(r.error)
        c.close()
    req_samples.append((time.perf_counter() - start) * 1000)
raw_samples.sort()
req_samples.sort()
e.close()
print(f"py raw={int(raw_samples[rounds // 2])} ns/render "
      f"request-loop={int(req_samples[rounds // 2])} ms/1000 rounds={rounds}")

File: bindings/node/bench/bench.js
// CP18 part B: Node binding raw render + repeated/server render workload.
const { Engine, Context } = require("../lib/nift.js");
const { performance } = require("perf_hooks");
(async () => {
  const e = Engine.new();
  e.setRoot("/");
  e.setString("site", "nift");
  const page = "<p>$[site]</p>";
  const tpl = "<main>@content</main>";
  const n = 50000;
  const rounds = 3;
  await e.render(page, tpl); // warm-up (unreported)
  const rawSamples = [], reqSamples = [];
  for (let r = 0; r < rounds; r++) {
    let start = performance.now();
    for (let i = 0; i < n; i++) { // raw: no request Context, engine-default binding
      const res = await e.render(page, tpl);
      if (!res.ok) throw new Error(res.error);
    }
    rawSamples.push((performance.now() - start) * 1e6 / n);
    start = performance.now();
    for (let i = 0; i < 1000; i++) { // request-loop: fresh Context per request
      const c = new Context();
      c.setString("who", "w");
      const res = await e.render(page, tpl, c);
      if (!res.ok) throw new Error(res.error);
      c.close();
    }
    reqSamples.push(performance.now() - start);
  }
  rawSamples.sort((a, b) => a - b);
  reqSamples.sort((a, b) => a - b);
  e.close();
  console.log(`node raw=${Math.round(rawSamples[1])} ns/render request-loop=${Math.round(reqSamples[1])} ms/1000 rounds=${rounds}`);
})();
